File: ios/Classes/FhSdkPlugin.m
#import "FhSdkPlugin.h"

#import <Flutter/Flutter.h>

#import <FH/FH.h>
#import <FH/FHResponse.h>

NSString * const PushMessageReceived = @"push_message_received";
NSString * const PushRegistrationSuccess = @"push_registration_success";
NSString * const PushRegistrationError = @"push_registration_error";

NSString * const DeviceTokenNotification = @"DeviceTokenNotification";

@interface FhSdkPlugin ()
@property(nonatomic, retain) FlutterMethodChannel *channel;
@end

@implementation FhSdkPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"fh_sdk"
      binaryMessenger:[registrar messenger]];
  FhSdkPlugin* instance = [[FhSdkPlugin alloc] init];
  instance.channel = channel;
  [registrar addMethodCallDelegate: instance channel: channel];
}

// Remove the observer once the instance is deallocated.
- (void)dealloc {
    //[self.channel setMethodCallHandler:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  // FH.init()
  if ([@"init" isEqualToString:call.method]) {
    [self handleInitCall: result];
  }
  // FH.cloudHost()
  else if ([@"getCloudUrl" isEqualToString:call.method]) {
      [self handleGetCloudUrlCall: result];
  }
  // FH.cloud()
  else if ([@"cloud" isEqualToString:call.method]) {
    [self handleCloudCall: call result: result];
  }
  // FH.auth()
  else if ([@"auth" isEqualToString:call.method]) {
    [self handleAuthCall: call result: result];
  }
  // FH.pushRegister()
  else if ([@"pushRegister" isEqualToString:call.method]) {
      [self handlePushRegisterCall: call result: result];
  }
  // FH.setPushAlias()
  else if ([@"setPushAlias" isEqualToString:call.method]) {
      [self handleSetPushAliasCall: call result: result];
  }
  // FH.setPushCategories()
  else if ([@"setPushCategories" isEqualToString:call.method]) {
      [self handleSetPushCategoriesCall: call result: result];
  }
  // Else...
  else {
    result(FlutterMethodNotImplemented);
  }
}

-(void)handleInitCall:(FlutterResult)result {
  NSLog(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
  
  // Call a cloud side function when init finishes
  void (^success)(FHResponse *)=^(FHResponse * res) {
      // Initialisation is now complete, you can now make FHActRequest's
      NSLog(@"SDK init = OK");
      result(@"SUCCESS");
  };
  
  void (^failure)(id)=^(FHResponse * res){
      NSString *errorMessage = [NSString stringWithFormat:@"Error: %@", res.rawResponseAsString];
      NSLog(@"init call exception. Status = %d Response = %@", res.responseStatusCode, errorMessage);
      result([FlutterError errorWithCode:@"INIT_ERROR"
                           message:errorMessage
                           details:res.parsedResponse]);
  };

  [FH initWithSuccess:success AndFailure:failure];
}

-(void)handleGetCloudUrlCall:(FlutterResult)result {
  NSLog(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    
  @try {
      NSString *cloudAppHost = [FH getCloudHost];
      NSLog(@"getCloudUrl: %@", cloudAppHost);
      result(cloudAppHost);
  }
  @catch ( NSException *e ) {
      NSString *errorMessage = [NSString stringWithFormat:@"Error: %@", [e reason]];
      NSLog(@"getCloudUrl call exception. Response = %@", errorMessage);
      result([FlutterError errorWithCode:@"CLOUDURL_ERROR"
                                 message:errorMessage
                                 details:[e reason]]);
  }
  @finally {
      NSLog(@"getCloudUrl finally area reached");
  }
}

-(void)handleCloudCall: (FlutterMethodCall*)call result:(FlutterResult)result {
  NSDictionary *options = call.arguments;
    
  NSLog(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
  NSLog(@"options: %@", options);
  
  // Call a cloud side function when init finishes
  void (^success)(FHResponse *)=^(FHResponse * res) {
      NSLog(@"cloud call succeded");
      result(res.rawResponseAsString);
  };
  
  void (^failure)(id)=^(FHResponse * res){
      NSDictionary *details = [[NSDictionary alloc] initWithObjectsAndKeys:
                               [NSNumber numberWithInt:res.responseStatusCode], @"statusCode",
                               res.rawResponseAsString, @"rawResponseAsString", nil];
      NSString *errorMessage = [NSString stringWithFormat:@"Error: %@", res.rawResponseAsString];
      NSLog(@"init call exception. Status = %d Response = %@", res.responseStatusCode, errorMessage);
      result([FlutterError errorWithCode:@"CLOUD_ERROR"
                                 message:errorMessage
                                 details:details]);
  };
  
  @try {
      NSString *path = nil, *method = nil, *contentType = nil;
      NSNumber *timeout = [NSNumber numberWithInteger:25000]; // 25 sec default
      NSMutableDictionary *headers = [[NSMutableDictionary alloc] init];
      NSMutableDictionary *data = [[NSMutableDictionary alloc] init];
      if (options && [options valueForKey:@"path"]) path = [options valueForKey:@"path"];
      if (options && [options valueForKey:@"method"]) method = [options valueForKey:@"method"];
      if (options && [options valueForKey:@"contentType"]) {
          contentType = [options valueForKey:@"contentType"];
      } else {
          contentType = @"application/json";
      }
      if (options && [options valueForKey:@"timeout"]) timeout = [options valueForKey:@"timeout"];
      
      if (options && [options valueForKey:@"headers"]) {
          [headers addEntriesFromDictionary: (NSDictionary*)[options valueForKey:@"headers"]];
      }
      [headers setValue:contentType forKey:@"contentType"];
      
      if (options && [options valueForKey:@"data"]) {
          [data addEntriesFromDictionary: (NSDictionary*)[options valueForKey:@"data"]];
      }
      
      FHCloudRequest * action = (FHCloudRequest *) [FH buildCloudRequest:path
                                                              WithMethod:method
                                                              AndHeaders:headers
                                                                  AndArgs:data];
      
      // change timeout (default value: 60s)
      action.requestTimeout = [timeout doubleValue]/1000;
      [action execAsyncWithSuccess: success AndFailure: failure];
  }
  @catch ( NSException *e ) {
      NSString *errorMessage = [NSString stringWithFormat:@"Error: %@", [e reason]];
      NSLog(@"cloud call exception. Response = %@", errorMessage);
      result([FlutterError errorWithCode:@"CLOUD_ERROR"
                                message:errorMessage
                                details:nil]);
  }
  @finally {
      NSLog(@"finally area reached");
  }
}

-(void)handleAuthCall: (FlutterMethodCall*)call result:(FlutterResult)result {
    NSLog(@"auth call with promise %@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    NSDictionary *options = call.arguments;
    
    NSString *authPolicy = [options valueForKey:@"authPolicy"];
    NSString *username = [options valueForKey:@"username"];
    NSString *password = [options valueForKey:@"password"];
    
    // Call a cloud side function when init finishes
    void (^success)(FHResponse *)=^(FHResponse * res) {
        NSLog(@"auth call succeded check status");
        NSLog(@"parsed response %@ type=%@",res.parsedResponse,[res.parsedResponse class]);
        if ([res responseStatusCode] != 200) {
            NSDictionary *details = [[NSDictionary alloc] initWithObjectsAndKeys:
                                     [NSNumber numberWithInt:res.responseStatusCode], @"statusCode",
                                     res.rawResponseAsString, @"rawResponseAsString", nil];
            NSString *errorMessage = [NSString stringWithFormat:@"Error: %@", res.rawResponseAsString];
            NSLog(@"auth call authentication failed. Status = %d Response = %@", res.responseStatusCode, errorMessage);
            result([FlutterError errorWithCode:@"AUTH_ERROR"
                                       message:errorMessage
                                       details:details]);
            
        } else {
            NSLog(@"auth call authentication succeded");
            result(res.rawResponseAsString);
        }
    };
    
    void (^failure)(id)=^(FHResponse * res){
        NSDictionary *details = [[NSDictionary alloc] initWithObjectsAndKeys:
                                 //[NSNumber numberWithInt:res.responseStatusCode], @"statusCode",
                                 res.rawResponseAsString, @"rawResponseAsString", nil];
        NSString *errorMessage = [NSString stringWithFormat:@"Error: %@", res.rawResponseAsString];
        NSLog(@"auth call exception. Status = %d Response = %@", res.responseStatusCode, errorMessage);
        result([FlutterError errorWithCode:@"AUTH_ERROR"
                                   message:errorMessage
                                   details:details]);
    };
    
    @try {
        
        if (!authPolicy || !username || !password) {
            NSString *errorMessage = [NSString stringWithFormat:@"Error authPolicy, username, password cannot empty"];
            result([FlutterError errorWithCode:@"AUTH_ERROR"
                                       message:errorMessage
                                       details:nil]);
            return;
        }
        
        
        NSLog(@"Policy: %@ username: %@", authPolicy, username);
        
        
        FHAuthRequest* authRequest = [FH buildAuthRequest];
        [authRequest authWithPolicyId:authPolicy UserId:username Password:password];
        
        [authRequest execAsyncWithSuccess:success AndFailure:failure];
    }
    @catch ( NSException *e ) {
        NSString *errorMessage = [NSString stringWithFormat:@"Error: %@", [e reason]];
        NSLog(@"cloud call exception. Response = %@", errorMessage);
        result([FlutterError errorWithCode:@"AUTH_ERROR"
                                   message:errorMessage
                                   details:nil]);
        
    }
    @finally {
        NSLog(@"finally area reached");
    }
}

-(void)handlePushRegisterCall: (FlutterMethodCall*)call result:(FlutterResult)result {
    NSLog(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    
    NSLog(@"pushRegister call %@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    NSDictionary *arguments = call.arguments;
    
    NSString *alias = [arguments valueForKey:@"alias"];
    NSArray *categories = [arguments valueForKey:@"categories"];
    
    if (!alias || !categories) {
        NSString *errorMessage = @"Neither alias nor categories can be null (categories can be empty though)";
        NSLog(@"pushRegister call exception. %@", errorMessage);
        result([FlutterError errorWithCode:@"PUSH_ERROR"
                                   message:errorMessage
                                   details:nil]);
    }
        
    // Call a cloud side function when init finishes
    void (^success)(FHResponse *)=^(FHResponse * res) {
        // Listen to DeviceTokenNotification
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(pushMessageReceived:)
                                                     name:PushMessageReceived
                                                   object:nil];
        
        // Initialisation is now complete, you can now make FHActRequest's
        NSLog(@"pushRegister call succeded");
        result(@"SUCCESS");
    };
    
    void (^failure)(id)=^(FHResponse * res){
        NSString *errorMessage = [NSString stringWithFormat:@"Error: %@", res.rawResponseAsString];
        NSLog(@"pushRegister call exception. Status = %d Response = %@", res.responseStatusCode, errorMessage);
        result([FlutterError errorWithCode:@"PUSH_ERROR"
                                   message:errorMessage
                                   details:res.parsedResponse]);
    };
    
    FHPushConfig *config = [[FHPushConfig alloc] init];
    config.alias = alias;
    config.categories = categories;
    [FH pushRegister:nil withPushConfig:config andSuccess:success andFailure:failure];
}

-(void)pushMessageReceived:(NSNotification *)notification {
    NSLog(@"%@ pushMessageReceived %@", NSStringFromClass([self class]), [notification description]);
    //[self.channel invokeMethod:@"push_message_received" arguments:[[notification userInfo] description]];
    NSDictionary *arguments = @{ @"object" : notification.object, @"userInfo" : notification.userInfo };
    [self.channel invokeMethod:@"push_message_received" arguments:arguments];
}

-(void)handleSetPushAliasCall: (FlutterMethodCall*)call result:(FlutterResult)result {
    NSLog(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    
    NSLog(@"setPushAlias call %@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    NSDictionary *arguments = call.arguments;
    
    NSString *alias = [arguments valueForKey:@"alias"];
    
    /*if (!alias) {
        NSString *errorMessage = @"Alias cannot be null";
        NSLog(@"setPushAlias call exception. %@", errorMessage);
        result([FlutterError errorWithCode:@"PUSH_ERROR"
                                   message:errorMessage
                                   details:nil]);
    }*/
    
    // Call a cloud side function when init finishes
    void (^success)(FHResponse *)=^(FHResponse * res) {
        // Listen to DeviceTokenNotification
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(pushMessageReceived:)
                                                     name:PushMessageReceived
                                                   object:nil];
        
        // Initialisation is now complete, you can now make FHActRequest's
        NSLog(@"setPushAlias call succeded");
        result(@"SUCCESS");
    };
    
    void (^failure)(id)=^(FHResponse * res){
        NSString *errorMessage = [NSString stringWithFormat:@"Error: %@", res.rawResponseAsString];
        NSLog(@"setPushAlias call exception. Status = %d Response = %@", res.responseStatusCode, errorMessage);
        result([FlutterError errorWithCode:@"PUSH_ERROR"
                                   message:errorMessage
                                   details:res.parsedResponse]);
    };
    
    [FH setPushAlias:alias andSuccess:success andFailure:failure];
}

-(void)handleSetPushCategoriesCall: (FlutterMethodCall*)call result:(FlutterResult)result {
    NSLog(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    
    NSLog(@"setPushCategories call %@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    NSDictionary *arguments = call.arguments;
    
    NSArray *categories = [arguments valueForKey:@"categories"];
    
    /*if (!categories) {
        NSString *errorMessage = @"Categories cannot be null but can be empty";
        NSLog(@"setPushCategories call exception. %@", errorMessage);
        result([FlutterError errorWithCode:@"PUSH_ERROR"
                                   message:errorMessage
                                   details:nil]);
    }*/
    
    // Call a cloud side function when init finishes
    void (^success)(FHResponse *)=^(FHResponse * res) {
        // Listen to DeviceTokenNotification
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(pushMessageReceived:)
                                                     name:PushMessageReceived
                                                   object:nil];
        
        // Initialisation is now complete, you can now make FHActRequest's
        NSLog(@"setPushCategories call succeded");
        result(@"SUCCESS");
    };
    
    void (^failure)(id)=^(FHResponse * res){
        NSString *errorMessage = [NSString stringWithFormat:@"Error: %@", res.rawResponseAsString];
        NSLog(@"setPushCategories call exception. Status = %d Response = %@", res.responseStatusCode, errorMessage);
        result([FlutterError errorWithCode:@"PUSH_ERROR"
                                   message:errorMessage
                                   details:res.parsedResponse]);
    };
    
    [FH setPushCategories:categories andSuccess:success andFailure:failure];
}

@end

File: android/src/main/java/com/redhat/fhsdk/FhSdkPlugin.java
package com.redhat.fhsdk;

import android.content.Context;

import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.PluginRegistry.Registrar;

import com.feedhenry.sdk.FH;
import com.feedhenry.sdk.FHActCallback;
import com.feedhenry.sdk.FHResponse;
import com.feedhenry.sdk.PushConfig;
import com.feedhenry.sdk.api.FHAuthRequest;
import com.feedhenry.sdk.api.FHCloudRequest;

import org.json.fh.JSONObject;

import cz.msebera.android.httpclient.Header;
import cz.msebera.android.httpclient.message.BasicHeader;


import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/**
 * FhSdkPlugin
 */
public class FhSdkPlugin implements MethodCallHandler {
    private final PluginRegistry.Registrar registrar;

    public FhSdkPlugin(Registrar registrar) {
        this.registrar = registrar;
    }

    /**
     * Plugin registration.
     */
    public static void registerWith(Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "fh_sdk");
        channel.setMethodCallHandler(new FhSdkPlugin(registrar));
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        // FH.init()
        if (call.method.equals("init")) {
            this.handleInitCall(result);
        }
        // FH.cloudHost()
        else if (call.method.equals("getCloudUrl")) {
            this.handleGetClourUrlCall(result);
        }
        // FH.cloud()
        else if (call.method.equals("cloud")) {
            this.handleCloudCall(call, result);
        }
        // FH.auth()
        else if (call.method.equals("auth")) {
            this.handleAuthCall(call, result);
        }
        // FH.pushRegister()
        else if (call.method.equals("pushRegister")) {
            this.handlePushRegisterCall(call, result);
        }
        else {
            result.notImplemented();
        }
    }

    private void handleInitCall (final Result result) {
        Context context = registrar.context();
        try {

            FH.init(context, new FHActCallback() {
                public void success(FHResponse pRes) {
                    // Initialisation is now complete, you can now make FHActRequest's
                    System.out.println("SDK initialised OK");
                    result.success("SUCCESS");
                }

                public void fail(FHResponse res) {
                    String errorMessage = "Error: " + res.getRawResponse();
                    System.out.println("init call exception. Response = " + res.getErrorMessage());
                    result.error("INIT_ERROR", errorMessage, res.getJson());
                }
            });
        } catch (Throwable e) {
            String errorMessage = "Exception: " + e.getMessage();
            System.out.println("init call exception. Response = " + errorMessage);
            result.error("INIT_ERROR", errorMessage, e.getCause());
        }
    }

    private void handleGetClourUrlCall (final Result result) {
        try {
            String cloudHost = FH.getCloudHost();
            System.out.println("CLOUD_URL cloudHost ==> " + cloudHost);
            result.success(cloudHost);

        } catch (Throwable e) {
            String errorMessage = "Exception: " + e.getMessage();
            System.out.println("getCloudUrl call exception. Response = " + errorMessage);
            result.error("CLOUD_URL_ERROR", errorMessage, e.getCause());
        }
    }

    private void handleCloudCall (final MethodCall call, final Result result) {
        HashMap options = (HashMap) call.arguments;

        if (options == null) {
            String errorMessage = "Wrong arguments options === null";
            System.out.println("cloud call exception. Response = " + errorMessage);
            result.error("CLOUD_ERROR", errorMessage, null);
            return;
        }

        String path = options.containsKey("path") ? (String) options.get("path") : null;
        String method = options.containsKey("method") ? (String) options.get("method") : null;
        String contentType = options.containsKey("contentType") ? (String) options.get("contentType") : "application/json";
        Integer timeout = null;
        try {
            timeout = options.containsKey("timeout") ? (Integer) options.get("timeout") : null;
        } catch (Throwable t) {
            String errorMessage = "Wrong arguments options.timeout format exception";
            System.out.println("cloud call exception. Response = " + errorMessage);
            result.error("CLOUD_ERROR", errorMessage, null);
            return;
        }

        Header[] headers = null;
        if (options.containsKey("headers")) {
            HashMap<String, String> _headers = (HashMap<String, String>) options.get("headers");
            if (_headers.size() > 0) {
                headers = new BasicHeader[_headers.size()];
                int i = 0;
                for (String key : _headers.keySet()) {
                    headers[i++] = new BasicHeader(key, _headers.get(key));
                }
            }
        }

        JSONObject data = null;
        if (options.containsKey("data")) {
            data = new JSONObject((Map)options.get("data"));
        }

        try {
            // Build the request object with request path, method, headers and data
            FHCloudRequest request = FH.buildCloudRequest(path, method, headers, data);
            // The request will be executed asynchronously
            request.executeAsync(new FHActCallback() {
                @Override
                public void success(FHResponse res) {
                    System.out.println("CLOUD res.getRawResponse ==> " + res.getRawResponse());
                    result.success(res.getRawResponse());
                    return;
                }

                @Override
                public void fail(FHResponse res) {
                    String errorMessage = "Error: " + res.getRawResponse();
                    System.out.println("cloud call exception. Response = " + res.getErrorMessage());
                    result.error("CLOUD_ERROR", errorMessage, res.getJson());
                }
            });
        } catch (Throwable e) {
            String errorMessage = "Exception: " + e.getMessage();
            System.out.println("cloud call exception. Response = " + errorMessage);
            result.error("CLOUD_ERROR", errorMessage, e.getCause());
        }
    }


    private void handleAuthCall (final MethodCall call, final Result result) {
        HashMap arguments = (HashMap) call.arguments;

        if (arguments == null) {
            String errorMessage = "Wrong arguments === null";
            System.out.println("cloud call exception. Response = " + errorMessage);
            result.error("CLOUD_ERROR", errorMessage, null);
            return;
        }

        String authPolicy = arguments.containsKey("authPolicy") ? (String) arguments.get("authPolicy") : null;
        String username = arguments.containsKey("username") ? (String) arguments.get("username") : null;
        String password = arguments.containsKey("password") ? (String) arguments.get("password") : null;

        try {
            // Build the auth object with request path, method, headers and data
            FHAuthRequest authRequest = FH.buildAuthRequest(authPolicy, username, password);
            authRequest.executeAsync(new FHActCallback() {
                @Override
                public void success(FHResponse res) {
                    System.out.println("AUTH res.getRawResponse ==> " + res.getRawResponse());
                    result.success(res.getRawResponse());
                    return;
                }

                @Override
                public void fail(FHResponse res) {
                    HashMap<String, String> details = new HashMap<>();
                    // TODO: find if this is possible... it is in iOS
                    //details.put("statusCode", "500");
                    details.put("rawResponseAsString", res.getRawResponse());
                    String errorMessage = "Error: " + res.getRawResponse();
                    System.out.println("auth call exception. Response = " + res.getErrorMessage());
                    result.error("AUTH_ERROR", errorMessage, details);
                }
            });
        } catch (Throwable e) {
            String errorMessage = "Exception: " + e.getMessage();
            System.out.println("init call exception. Response = " + errorMessage);
            result.error("CLOUD_ERROR", errorMessage, e.getCause());
        }
    }

    private void handlePushRegisterCall (final MethodCall call, final Result result) {
        HashMap arguments = (HashMap) call.arguments;

        PushConfig config = null;
        String alias = null;
        ArrayList<String> categories = null;

        try {
            if (arguments != null) {
                config = new PushConfig();
                config.setAlias(arguments.containsKey("alias") ? (String) arguments.get("alias") : null);
                config.setCategories(arguments.containsKey("categories") ? (ArrayList<String>) arguments.get("categories") : null);
            }

            FHActCallback callback = new FHActCallback() {
                public void success(FHResponse res) {
                    // Initialisation is now complete, you can now make FHActRequest's
                    System.out.println("PUSH res.getRawResponse ==> " + res.getRawResponse());
                    result.success(res.getRawResponse());
                }

                @Override
                public void fail(FHResponse res) {
                    String errorMessage = "Error: " + res.getRawResponse();
                    System.out.println("pushRegister call exception. Response = " + res.getErrorMessage());
                    result.error("PUSH_ERROR", errorMessage, res.getJson());
                }
            };
            if (config != null) {
                FH.pushRegister(config, callback);
            } else {
                FH.pushRegister(callback);
            }
        } catch (Throwable e) {
            String errorMessage = "Exception: " + e.getMessage();
            System.out.println("pushRegister call exception (check alias and categories types if in use). Response = " + errorMessage);
            result.error("PUSH_ERROR", errorMessage, e.getCause());
        }
    }


}
